README.md
# Fast Step Graph

![image](https://github.com/juancolonna/FastStepGraph/assets/6243522/2c0f3f9c-e675-4bd9-bd3e-0f4ff197efdb)

Fast Step Graph is an optimized implementation in **R** of the Stepwise approach used for discovering high-dimensional Gaussian Graphical Models. It aims to accurately estimate the $\mathbf{\Omega}$ precision matrix when dealing with datasets where the number of features is significantly larger than the number of samples ($p >> n$), such as in genomics.

This implementation builds upon the original code available at [link](https://jdssv.org/index.php/jdssv/article/view/11), which accompanies the associated research paper. **Fast Step Graph** enhances the computational efficiency of the original code to handle much larger graph structures than previously reported, while reducing the training time. Several improvements have been made, including the elimination of redundant code, utilization of column-wise data structures (better for R), avoid list creation, manipulation and expansion within loops, and the integration of a faster subroutine for regression. Additionally, this implementation addresses a bug introduced in the original code.

Despite these enhancements, the primary bottleneck of Fast Step Graph lies in the requirement of substantial memory resources $\text{Memory} \propto \Big(\frac{p(p-1)}{2}\Big)$ for storing the entire graph, particularly when $p$ grows.

Clone this repository or simply download the .zip file and follow the instructions in [this link to see an example](https://github.com/juancolonna/FastStepGraph/blob/main/vignettes/How_to_use.pdf).

How to cite this repository?

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.8233706.svg)](https://doi.org/10.5281/zenodo.8233706)


NEWS.md
Changes in version 0.1.2

Improvements

    - We added a new checking routine in cv.FastStepGraph that prevents the generation of incorrect results when for some reason the input data cannot be adjusted with cross-validation.
    
Changes in version 0.1.1

Improvements

    - We added a new parameter called 'return_model' to the cv.FastStepGraph function to specify when to return the fitted model, i.e. internally calls FastStepGraph with the optimal alpha_f and alpha_b parameters and returns Omega.

    - We added a new parameter called b_coef with initial value 0.5. This parameter serves to control the empirical rule alpha_b=b_coef*alpha_f during the initial search for the optimal alpha_f parameter while alpha_b remains fixed, after finding optimal alpha_f, alpha_b is varied to find its ideal value. The default value of b_coef is 0.5, but with real data sets it may be necessary to use a lower value.

Bug fixes

    - In the previous version only one value of alpha_b was tested, now several different values of this parameter are tested to find the ideal value that produces the minimum loss during cross-validation.

# FastStepGraph 0.1.0

* Initial CRAN submission.


R/SigmaAR.R
#' @title Simulate Covariance Matrix with an Auto-regressive (AR) Model
#'
#' @description Helper function to simulate Simulate Gaussian Data with an Autoregressive (AR) Model
#'
#' @param n_rows Number of samples (rows of X).
#' @param p_columns Number of variables (columns of X).
#' @param phi Auto-regression coefficient.
#'
#' @return A list with the values:
#' \item{\code{Sigma}}{A covariance matrix.}
#' \item{\code{Omega}}{A precision matrix.}
#' \item{\code{X}}{A normalized data matrix with Gaussian distribution.}
#'
#' @importFrom MASS mvrnorm
#' 
#' @author Prof. Juan G. Colonna, PhD. \email{juancolonna@icomp.ufam.edu.br}
#' @author Prof. Marcelo Ruiz, PhD. \email{mruiz@exa.unrc.edu.ar}
#'
#' @export
#' 
#' @examples
#' data <- FastStepGraph::SigmaAR(30, 50, 0.4) # Simulate Gaussian Data
SigmaAR <- function(n_rows, p_columns, phi){
    Sigma <- diag(p_columns)
    for (i in seq_len(p_columns)) {
        for (j in seq_len(p_columns)) {
            Sigma[i,j] <- phi^(abs(i-j))
        }
    }
    
    Omega <- solve(Sigma)
    Omega[abs(Omega) < 1e-5] <- 0
    X <- list() # Generate Data from a Gaussian distribution
    X <- MASS::mvrnorm(n_rows, mu=rep(0, p_columns), Sigma)
    X <- scale(X) # data normalization to zero mean and unit variance 
    return(list(Sigma=Sigma, Omega=Omega, X=X))
}



R/FastStepGraph.R
#' @title Fast Stepwise Gaussian Graphical Model
#'
#' @description Improved and faster implementation of the Stepwise Gaussian Graphical Algorithm.
#'
#' @param x Data matrix (of size n_samples x p_variables).
#' @param alpha_f Forward threshold  (no default value).
#' @param alpha_b Backward threshold. If alpha_b=NULL, then the rule alpha_b <- 0.5*alpha_f is applied.
#' @param nei.max Maximum number of variables in every neighborhood (default value 5).
#' @param data_scale Boolean parameter (TRUE or FALSE), when to scale data to zero mean and unit variance (default FALSE).
#' @param max.iterations Maximum number of iterations (integer), the defaults values is set to p*(p-1).
#'
#' @return A list with the values:
#' \item{\code{vareps}}{Response variables.}
#' \item{\code{beta}}{Regression coefficients.}
#' \item{\code{Edges}}{Estimated set of edges.}
#' \item{\code{Omega}}{Estimated precision matrix.}
#'
#' @author Prof. Juan G. Colonna, PhD. \email{juancolonna@icomp.ufam.edu.br}
#' @author Prof. Marcelo Ruiz, PhD. \email{mruiz@exa.unrc.edu.ar}
#'
#' @export
#' 
#' @examples
#' data <- FastStepGraph::SigmaAR(30, 50, 0.4) # Simulate Gaussian Data
#' G <- FastStepGraph::FastStepGraph(data$X, alpha_f = 0.22, alpha_b = 0.14, data_scale=TRUE)
FastStepGraph <- function(x, 
                          alpha_f, 
                          alpha_b=NULL, 
                          nei.max=5, 
                          data_scale=FALSE, 
                          max.iterations=NULL){
  
  .lm.fit = combn = cor = cov = NULL
  if (data_scale) { x <- scale(x) }
  n <- dim(x)[1] # number of rows
  p <- dim(x)[2] # number of columns

  if (alpha_f < alpha_b){ stop("alpha_b must be lower than alpha_f") }
  if (is.null(alpha_b)){ alpha_b <- 0.5*alpha_f }
  if (nei.max == 0) { stop('The minimum number of neighbors (nei.max) must be greater than 0.') }
  if (nei.max >= n && n <= p) { stop('Neiborgs must be less than n-1') }
  if (nei.max >= p && p <= n) { stop('Neiborgs must be less than p-1') }

  Edges_I <- t(combn(seq_len(p),2)) # Inactive set of ordered pair (i,j)
  Edges_A <- t(matrix(0,2,dim(Edges_I)[1])) # Active set of ordered pair (i,j)
  N_neighbors <- integer(p) # number of neighbors of each node

  e <- x # (n x p) matrix of regression residuals
  f_ij <- cor(e)
  if (any(is.na(f_ij))) { stop('Columns with zero variance found in input data') }
  lower_tri <- lower.tri(f_ij)
  f_ij <- abs(f_ij[lower_tri])
  b_ij <- rep(2, length(f_ij))

  k <- 1
  if (is.null(max.iterations)){ max.iterations <- p*(p-1) } # length(Edges_I[,2]) is the total number of edges

  while ((k <= max.iterations)) {
    # Select (i,j) that max(|f_ij|)
    f_ij_indx <- which.max(f_ij)
    f_ij_max <- f_ij[f_ij_indx]

    i_f <- Edges_I[f_ij_indx, 1]
    j_f <- Edges_I[f_ij_indx, 2]

    if ( f_ij_max < alpha_f ) {break} # if |f_ij| lower than alpha, then stop.

    # If the number of neighbors for a given node exceeds the maximum allowed number
    # of neighbors (nei.max), refrain from adding a new neighbor to that node.
    else if ((N_neighbors[i_f]+1) > nei.max || (N_neighbors[j_f]+1) > nei.max) { f_ij[f_ij_indx] <- 0 }

    else{
      ################ Forward-Step ######################################
      N_neighbors[i_f] <- N_neighbors[i_f]+1
      N_neighbors[j_f] <- N_neighbors[j_f]+1

      Edges_A[f_ij_indx,] <- Edges_I[f_ij_indx,]
      Edges_I[f_ij_indx,] <- c(0,0)

      # Update Prediction Errors for (i_f,j_f)
      n_i_f_pos_2 <- which(Edges_A[,2]==i_f)
      n_i_f_pos_1 <- which(Edges_A[,1]==i_f)
      n_i_f <- c(Edges_A[n_i_f_pos_2, 1], Edges_A[n_i_f_pos_1, 2])

      n_j_f_pos_2 <- which(Edges_A[,2]==j_f)
      n_j_f_pos_1 <- which(Edges_A[,1]==j_f)
      n_j_f <- c(Edges_A[n_j_f_pos_2, 1], Edges_A[n_j_f_pos_1, 2])

      # fast LM documentation https://rpubs.com/maechler/fast_lm
      e[,i_f] <- .lm.fit(cbind(1, x[,n_i_f]), x[,i_f])$residuals
      e[,j_f] <- .lm.fit(cbind(1, x[,n_j_f]), x[,j_f])$residuals

      ############### Backward-Step ######################################
      # Compute Prediction Errors r_i and r_j for (i,j) in the active set Edges_A
      L <- unique(c(n_i_f_pos_2, n_i_f_pos_1, n_j_f_pos_1, n_j_f_pos_2))
      b_ij[f_ij_indx] <- f_ij_max
      L <- L[-which(L == f_ij_indx)]

      ############### Residuals update ###################################
      for (l in L) {
        i <- Edges_A[l, 1]
        j <- Edges_A[l, 2]

        n_i <- c(Edges_A[which(Edges_A[,2] == i), 1], Edges_A[which(Edges_A[,1] == i), 2])
        n_i <- n_i[-which(n_i==j)]
        if(length(n_i) > 0) { r_i <- .lm.fit(cbind(1, x[,n_i]), x[,i])$residuals } 
        else { r_i <-  x[,i] }

        n_j <- c(Edges_A[which(Edges_A[,2] == j), 1], Edges_A[which(Edges_A[,1] == j), 2])
        n_j <- n_j[-which(n_j==i)]
        if(length(n_j) > 0) { r_j <- .lm.fit(cbind(1, x[,n_j]), x[,j])$residuals }
        else { r_j <- x[,j] }

        # b_ij[l] <- abs(cor(r_i, r_j))
        corr_b <- cor(r_i, r_j)
        b_ij[l] <- ifelse(is.na(corr_b), 0, abs(corr_b)) # I have to check this condition
      }

      ################ Edge elimination ######################################
      # Select (i,j) that min(|b_ij|) <= alpha_b
      b_ij_indx <- which.min(b_ij)
      b_ij_min <- b_ij[b_ij_indx]

      # Update active set of edges Edges_A
      if (b_ij_min <= alpha_b){

        i_b <- Edges_A[b_ij_indx, 1]
        j_b <- Edges_A[b_ij_indx, 2]

        N_neighbors[i_b] <- N_neighbors[i_b]-1
        N_neighbors[j_b] <- N_neighbors[j_b]-1

        Edges_I[b_ij_indx,] <- Edges_A[b_ij_indx,]
        Edges_A[b_ij_indx,] <- c(0,0)

        # Update Prediction Errors for (i_b,j_b)
        n_i_b <- c(Edges_A[which(Edges_A[,2] == i_b), 1], Edges_A[which(Edges_A[,1] == i_b), 2])
        n_j_b <- c(Edges_A[which(Edges_A[,2] == j_b), 1], Edges_A[which(Edges_A[,1] == j_b), 2])

        if(N_neighbors[i_b] > 0) { e[,i_b] <- .lm.fit(cbind(1, x[,n_i_b]), x[,i_b])$residuals }
        else { e[,i_b] <- x[,i_b] }
        
        if(N_neighbors[j_b] > 0) { e[,j_b] <- .lm.fit(cbind(1, x[,n_j_b]), x[,j_b])$residuals }
        else { e[,j_b] <- x[,j_b] }

        b_ij[b_ij_indx] <- 2
      }

      f_ij <- cor(e)
      f_ij <- abs(f_ij[lower_tri])
      f_ij[is.na(f_ij)] <- 0 # I have to check this condition
      H <- which(Edges_A[,1]>0)
      f_ij[H] <- 0


      ############# steps of main loop (while) ###########################
      k <- k+1
      # if (k > max.iterations){ message('Maximum number iteration reached') }
    }
  }

  ################## Release some memory #####################################
  rm(f_ij, b_ij, Edges_I, N_neighbors)

  ################## Compute the precision matrix ############################
  col_var <- colvars(e)
  cor_matrix <- cov(e)
  cor_matrix[is.na(cor_matrix)] <- 0 # I have to check this condition
  Omega <- matrix(0,p,p)
  diag(Omega) <- col_var^(-1)
  Omega[is.infinite(Omega)] <- 1/.Machine$double.eps # I have to check this condition
  beta <- matrix(0,p,p)

  for (edge in which(Edges_A[,2] > 0)){
    i <- Edges_A[edge, 1]
    j <- Edges_A[edge, 2]

    Omega[i,j] <- cor_matrix[i,j]*Omega[i,i]*Omega[j,j]
    Omega[j,i] <- Omega[i,j]
    
    if(col_var[j] <= .Machine$double.eps) {beta[i,j] <- 0} # I have to check this condition
    else {beta[i,j] <- -cor_matrix[i,j]/col_var[j]}
    
    if(col_var[i] <= .Machine$double.eps) {beta[j,i] <- 0} # I have to check this condition
    else {beta[j,i] <- -cor_matrix[i,j]/col_var[i]}
  }
  return(list(vareps=e, beta=beta, Edges=Edges_A, Omega=Omega))
}

colvars <- function(x) {
  means <- colMeans(x)
  n <- length(x)/length(means)
  centeredX <- x - rep(means, each = n)
  colSums(centeredX^2) / (n-1)
}

grid.cv.FastStepGraph.R
#'
#' @title Searches for the optimal combination of alpha_f and alpha_b parameters using Cross-Validation
#'
#' @description \code{cv.FastStepGraph} implements the cross-valiation for the Fast Step Graph algorithm.
#'
#' @param x Data matrix (of size n x p).
#' @param n_folds Number of folds for the cross-validation procedure (default value 5).
#' @param alpha_f_min Minimum alpha_f threshold value for the cross-validation procedure (default value 0.1).
#' @param alpha_f_max Maximum alpha_f threshold value for the cross-validation procedure (default value 0.9).
#' @param n_alpha Number of elements in the grid for the cross-validation (default value 32).
#' @param nei.max Maximum number of variables in every neighborhood (default value 5).
#' @param data_scale Boolean parameter (TRUE or FALSE), when to scale data to zero mean and unit variance (default FALSE).
#' @param return_model Boolean parameter (TRUE or FALSE), when to return the output of FastStepGraph function (default TRUE).
#' @param parallel Boolean parameter (TRUE or FALSE), when to run Cross-Validation in parallel using a multicore architecture (default FALSE).
#'
#' @return A list with the values: \cr \cr
#' \item{\code{alpha_f_opt}: the optimal alpha_f value.}
#' \item{\code{alpha_f_opt}: the optimal alpha_f value.}
#' \item{\code{CV.loss}: minimum loss.}
#' \item{\code{Loss_grid}: a loss grid.} 
#' \item{\code{vareps}: Response variables.}
#' \item{\code{beta}: Regression coefficients.}
#' \item{\code{Edges}: Estimated set of edges.}
#' \item{\code{Omega}: Estimated precision matrix.}
#'
#' @export
#'
#' @author Prof. Juan G. Colonna, PhD. \email{juancolonna@icomp.ufam.edu.br}
#'

grid.cv.FastStepGraph = function(x, n_folds=5, alpha_f_min=0.1, alpha_f_max=0.9, n_alpha=32, nei.max = 5, data_scale = FALSE, parallel = FALSE){
    n = nrow(x)
    p = ncol(x)
    
    if (n_folds <= 1) { stop('Number of folds must be equal or larger than 2') }
    if (nei.max >= n) { stop('The maximum number of neighbors (nei.max) must be less than n-1.') }
    if (nei.max == 0) { stop('The minimum number of neighbors (nei.max) must be greater than 0.') }
    if (data_scale) { x = scale(x) }

    ntest = floor(n/n_folds)
    ntrain = n - ntest
    ind = sample(n)
    alpha_f = seq(alpha_f_min, alpha_f_max, length=n_alpha)
    alpha_b = alpha_f
    loss_grid = matrix(NA, nrow = n_alpha, ncol = n_alpha)

    old_loss = Inf
    new_loss = NA
    alpha_f_opt = NA
    alpha_b_opt = NA
    
    if (parallel) { 
      if(.Platform$OS.type == "unix") {
        library(doParallel) # needed to parallelize on Linux
        cores = detectCores()
        cl = makeCluster(cores[1]-1, type = "FORK") # not to overload your computer "FORK"
        registerDoParallel(cl)
      } 
      else {
        library(doSNOW) # needed to parallelize on Windows
        cores = detectCores()
        cl = makeCluster(cores[1]-1, type="SOCK")
        registerDoSNOW(cl)
      }
      
      losses_and_alphas <- foreach(i = 1:n_alpha, .combine = 'c', .inorder=TRUE) %dopar% {
      source('FastStepGraph.R')

          for (j in 1:n_alpha) {
              if (alpha_f[i] >= alpha_b[j]) {
                  loss = 0
                  for (k in 1:n_folds) {
                      sel = ((k-1)*ntest+1):(k*ntest)
                      x.train = x[ind[-sel], ]
                      x.test = x[ind[sel], ]
                      beta = FastStepGraph(x.train,
                                           alpha_f = alpha_f[i],
                                           alpha_b = alpha_b[j],
                                           nei.max = nei.max)$beta
                      loss = loss + sum(colSums((x.test - x.test%*%beta)^2))
                  }
                  new_loss = loss/n_folds
                  loss_grid[i, j] = new_loss
                  if (new_loss < old_loss) {
                      old_loss = new_loss
                      alpha_f_opt = alpha_f[i]
                      alpha_b_opt = alpha_b[j]
                  }
              }
          }
        c(old_loss, alpha_f_opt, alpha_b_opt)
      }
      stopCluster(cl) #stop cluster

      best_loss = losses_and_alphas[1]
      alpha_f_opt = losses_and_alphas[2]
      alpha_b_opt = losses_and_alphas[3]
      for(i in seq(4, (3*n_alpha)-2, by = 3)){
        if (losses_and_alphas[i] < best_loss) {
          best_loss = losses_and_alphas[i]
          alpha_f_opt = losses_and_alphas[i+1]
          alpha_b_opt = losses_and_alphas[i+2]
        }
      }
    } # if (parallel)
    
    else {
      for (i in 1:n_alpha) {
        for (j in 1:n_alpha) {
          if (alpha_f[i] >= alpha_b[j]) {
            loss = 0
            for (k in 1:n_folds) {
              sel = ((k-1)*ntest+1):(k*ntest)
              x.train = x[ind[-sel], ]
              x.test = x[ind[sel], ]
              beta = FastStepGraph(x.train,
                                   alpha_f = alpha_f[i],
                                   alpha_b = alpha_b[j],
                                   nei.max = nei.max)$beta
              loss = loss + sum(colSums((x.test - x.test%*%beta)^2))
            }
            new_loss = loss/n_folds
            loss_grid[i, j] = new_loss
            if (new_loss < old_loss) {
              old_loss = new_loss
              alpha_f_opt = alpha_f[i]
              alpha_b_opt = alpha_b[j]
            }
          }
        }
      }
    }

    G = FastStepGraph(x, alpha_f = alpha_f_opt, alpha_b = alpha_b_opt, nei.max = nei.max)
    return(list(alpha_f_opt = alpha_f_opt, 
                alpha_b_opt = alpha_b_opt, 
                CV.loss = old_loss, 
                Loss_grid = loss_grid,
                vareps = G$vareps, 
                beta = G$beta, 
                Edges = G$Edges, 
                Omega = G$Omega))
 }


R/cv.FastStepGraph.R
#' @title Searches for the optimal combination of alpha_f and alpha_b parameters using Cross-Validation
#'
#' @description \code{cv.FastStepGraph} implements the cross-validation for the Fast Step Graph algorithm.
#'
#' @param x Data matrix (of size n x p).
#' @param alpha_f_min Minimum alpha_f value for the cross-validation procedure (example 0.1).
#' @param alpha_f_max Maximum alpha_f value for the cross-validation procedure  (example 0.9).
#' @param n_folds Number of folds for the cross-validation procedure (default value 10). This parameter also accepts the string 'LOOCV' to perform Leave-One-Out cross-validation.
#' @param b_coef This parameter applies the empirical rule alpha_b=b_coef*alpha_f during the initial search for the optimal alpha_f parameter while alpha_b remains fixed, after finding optimal alpha_f, alpha_b is varied to find its optimal value. The default value of b_coef is 0.5.
#' @param n_alpha Number of elements in the grid for the cross-validation (default value 20).
#' @param nei.max Maximum number of variables in every neighborhood (default value 5).
#' @param data_scale Boolean parameter (TRUE or FALSE), when to scale data to zero mean and unit variance (default FALSE).
#' @param data_shuffle Boolean parameter (default TRUE), when samples (rows of X) must be randomly shuffled.
#' @param max.iterations Maximum number of iterations (integer), the defaults values is set to p*(p-1).
#' @param return_model Default FALSE. If set to TRUE, at the end of cross-validation, FastStepGraph is called with the optimal parameters alpha_f and alpha_b, returning \code{vareps}, \code{beta}, \code{Edges} and \code{Omega}.
#' @param parallel Boolean parameter (TRUE or FALSE), when to run Cross-Validation in parallel using a multicore architecture (default FALSE).
#' @param n_cores An 'int' value specifying the number of cores do you want to use if 'parallel=TRUE'. 
#' If n_cores is not specified, the maximum number of cores on your machine minus one will be set automatically.
#' 
#' @return A list with the values:
#' \item{\code{alpha_f_opt}}{the optimal alpha_f value.}
#' \item{\code{alpha_f_opt}}{the optimal alpha_f value.}
#' \item{\code{CV.loss}}{minimum loss.}
#' If return_model=TRUE, then also returns:
#' \item{\code{vareps}}{Response variables.}
#' \item{\code{beta}}{Regression coefficients.}
#' \item{\code{Edges}}{Estimated set of edges.}
#' \item{\code{Omega}}{Estimated precision matrix.}
#'
#' @author Prof. Juan G. Colonna, PhD. \email{juancolonna@icomp.ufam.edu.br}
#' @author Prof. Marcelo Ruiz, PhD. \email{mruiz@exa.unrc.edu.ar}
#'
#' @export
#' @importFrom foreach %dopar%
#' 
#' @examples
#' data <- FastStepGraph::SigmaAR(30, 50, 0.4) # Simulate Gaussian Data
#' res <- FastStepGraph::cv.FastStepGraph(data$X, alpha_f_min=0.1, alpha_f_max = 0.9, data_scale=TRUE)
cv.FastStepGraph <- function(x, 
                             alpha_f_min, 
                             alpha_f_max, 
                             n_folds = 10,
                             b_coef = 0.5,
                             n_alpha = 20, 
                             nei.max = 5, 
                             data_scale = FALSE, 
                             data_shuffle = TRUE, 
                             max.iterations = NULL,
                             return_model = FALSE,
                             parallel = FALSE,
                             n_cores = NULL){
  f <- NULL
  n <- nrow(x)
  p <- ncol(x)

  if (alpha_f_max >= 1) { stop('Please, decrease alpha_f_max.') }
  if (alpha_f_min <= 0) { stop('Please, increase alpha_f_min.') }
  if (alpha_f_min >= alpha_f_max) { stop('alpha_f_min must be less than alpha_f_max.') }
  if (nei.max == 0) { stop('The minimum number of neighbors (nei.max) must be greater than 0.') }
  if (nei.max >= n && n <= p) { stop('Neiborgs must be less than n-1') }
  if (nei.max >= p && p <= n) { stop('Neiborgs must be less than p-1') }
  if (nei.max >= p && p <= n) { stop('Neiborgs must be less than p-1') }
  if (n_alpha <= 3) { stop('n_alpha must be larger than 3') }
  if (n_folds < 2) { stop('Number of folds must be equal or larger than 2') }
  if (is.numeric(n_cores)) { if (n_cores <= 0) {stop('n_cores must be greater than 0.')}}
  if (data_scale) { x <- scale(x) }
  if (data_shuffle) { x <- x[sample(seq_len(n)),] }
  if (is.null(max.iterations)){ max.iterations <- p*(p-1) }
  if (n_folds == 'LOOCV' ) { n_folds = n }
  
  ntest <- floor(n/n_folds)
  ntrain <- n - ntest
  alpha_f <- seq(alpha_f_min, alpha_f_max, length=n_alpha)

  min_loss <- Inf
  new_loss <- NA
  alpha_f_opt <- NA
  alpha_b_opt <- NA

  if (parallel) {
    if (is.null(n_cores)) { n_cores <- parallel::detectCores()-1 }  # -1 to not overload your computer 
    if(.Platform$OS.type == "unix") {
      cl <- parallel::makeCluster(n_cores[1], type = "FORK") # in linux "FORK"
      doParallel::registerDoParallel(cl)
    }
    else {
      cl <- parallel::makeCluster(n_cores[1], type="SOCK") # in windows "SOCK"
      doParallel::registerDoParallel(cl)
    }

    alpha_f_losses <- foreach::foreach(f = alpha_f, .combine = 'c', .inorder=TRUE) %dopar% {
      loss <- 0
      for (k in seq_len(n_folds)) {
        sel <- ((k-1)*ntest+1):(k*ntest)
        x.train <- x[-sel, ]
        x.test <- x[sel, ]
        beta <- FastStepGraph(x.train,
                             alpha_f = f,
                             alpha_b = b_coef*f,
                             nei.max = nei.max,
                             max.iterations = max.iterations)$beta
        loss <- loss + sum(colSums((x.test - x.test%*%beta)^2))
      }
      loss/n_folds
    }
    
    # avoiding NA
    alpha_f <- alpha_f[!is.na(alpha_f_losses)]
    alpha_f_losses <- alpha_f_losses[!is.na(alpha_f_losses)]

    if (length(alpha_f) == 0) { 
      parallel::stopCluster(cl)
      stop("Convergence failed, no alpha_f value found (possibly due to a problem in the input data).
            To solve this issue, try increasing alpha_f_min, decreasing b_coef, and/or increasing n_folds. 
            You can also add Gaussian noise to the data with zero mean and very small variance.") 
    }
    
    indx_min_loss <- min(which.min(alpha_f_losses))
    min_loss = alpha_f_losses[indx_min_loss]
    alpha_f_opt <- alpha_f[indx_min_loss]
    alpha_b_opt <- b_coef*alpha_f_opt

    alpha_b <- seq(0.1, 0.9*alpha_f_opt, length=10*alpha_f_opt)
    alpha_b_losses <- foreach::foreach(b = alpha_b, .combine = 'c', .inorder=TRUE) %dopar% {
      loss <- 0
      for (k in seq_len(n_folds)) {
        sel <- ((k-1)*ntest+1):(k*ntest)
        x.train <- x[-sel, ]
        x.test <- x[sel, ]
        beta <- FastStepGraph(x.train,
                             alpha_f = alpha_f_opt,
                             alpha_b = b,
                             nei.max = nei.max,
                             max.iterations = max.iterations)$beta
        loss <- loss + sum(colSums((x.test - x.test%*%beta)^2))
      }
      loss/n_folds
    }

    parallel::stopCluster(cl) #stop cluster
    
    # avoiding NA
    alpha_b <- alpha_b[!is.na(alpha_b_losses)]
    alpha_b_losses <- alpha_b_losses[!is.na(alpha_b_losses)]
    
    if (length(alpha_b) == 0) { }
    else if (min(alpha_b_losses) < min(alpha_f_losses)){
      indx_min_loss <- min(which.min(alpha_b_losses))
      min_loss = alpha_b_losses[indx_min_loss]
      alpha_b_opt <- alpha_b[indx_min_loss]
    }
  }
  
  # Not Parallel loop
  else {
    for (f in alpha_f) {
      loss <- 0
      for (k in seq_len(n_folds)) {
        sel <- ((k-1)*ntest+1):(k*ntest)
        x.train <- x[-sel, ]
        x.test <- x[sel, ]
        beta <- FastStepGraph(x.train,
                             alpha_f = f,
                             alpha_b = b_coef*f,
                             nei.max = nei.max,
                             max.iterations = max.iterations)$beta
        loss <- loss + sum(colSums((x.test - x.test%*%beta)^2))
      }
 
      if (is.na(loss)){ }
      else if (loss/n_folds < min_loss) {
        min_loss <- loss/n_folds
        alpha_f_opt <- f
        alpha_b_opt <- b_coef*f
      }
    }
    
    if (is.infinite(min_loss)) {
      stop("Convergence failed, no alpha_f value found (possibly due to a problem in the input data).
            To solve this issue, try increasing alpha_f_min, decreasing b_coef, and/or increasing n_folds. 
            You can also add Gaussian noise to the data with zero mean and very small variance.")
    }
    
    alpha_b <- seq(0.1, 0.9*alpha_f_opt, length=10*alpha_f_opt)
    for (b in alpha_b) {
      loss <- 0
      for (k in seq_len(n_folds)) {
        sel <- ((k-1)*ntest+1):(k*ntest)
        x.train <- x[-sel, ]
        x.test <- x[sel, ]
        beta <- FastStepGraph(x.train,
                             alpha_f = alpha_f_opt,
                             alpha_b = b,
                             nei.max = nei.max,
                             max.iterations = max.iterations)$beta
        loss <- loss + sum(colSums((x.test - x.test%*%beta)^2))
      }
      if (is.na(loss)){ }
      else if (loss/n_folds < min_loss) {
        min_loss <- loss/n_folds
        alpha_b_opt <- b
      }
    }
  }
  
  if (return_model) {
    G <- FastStepGraph(x, alpha_f = alpha_f_opt, alpha_b = alpha_b_opt, nei.max = nei.max)
   return(list(alpha_f_opt = alpha_f_opt,
              alpha_b_opt = alpha_b_opt,
              CV.loss = min_loss,
              vareps = G$vareps,
              beta = G$beta,
              Edges = G$Edges,
              Omega = G$Omega))
  }
  else {
    return(list(alpha_f_opt = alpha_f_opt,
                alpha_b_opt = alpha_b_opt,
                CV.loss = min_loss))
  }
}
